test/util/periods_unit.js

```diff
@@ -1141,14 +1149,163 @@ describe('PeriodCombiner', () => {
     expect(audio2.originalId).toBe('stream2,stream4');
   });
 
+  it('Matches streams with roles in common', async () => {
+    /** @type {!Array.<shaka.util.PeriodCombiner.Period>} */
+    const periods = [
+      {
+        id: '1',
+        videoStreams: [
+          makeVideoStream(720),
+          makeVideoStream(1080),
+        ],
+        audioStreams: [
+          makeAudioStreamWithRoles('stream1', ['main']),
+          makeAudioStreamWithRoles('stream2', ['description'], false),
+        ],
+        textStreams: [],
+        imageStreams: [],
+      },
+      {
+        id: '2',
+        videoStreams: [
+          makeVideoStream(720),
+          makeVideoStream(1080),
+        ],
+        audioStreams: [
+          makeAudioStreamWithRoles('stream1', ['main']),
+          makeAudioStreamWithRoles('stream2', ['description'], false),
+        ],
+        textStreams: [],
+        imageStreams: [],
+      },
+    ];
+
+    await combiner.combinePeriods(periods, /* isDynamic= */ false);
+    const variants = combiner.getVariants();
+
+    console.log(variants);
+
+    expect(variants.length).toBe(4);
+
+    expect(variants).toEqual(jasmine.arrayWithExactContents([
+      makeAVVariant(720, 'en'),
+      makeAVVariant(1080, 'en'),
+      makeAVVariant(720, 'en'),
+      makeAVVariant(1080, 'en'),
+    ]));
+
+    // We can use the originalId field to see what each track is composed of.
+    const audio1 = variants[0].audio;
+    expect(audio1.roles).toEqual(['main']);
+    expect(audio1.originalId).toBe('stream1,stream1');
+
+    const audio2 = variants[1].audio;
+    expect(audio2.roles).toEqual(['main']);
+    expect(audio2.originalId).toBe('stream1,stream1');
+
+    const audio3 = variants[2].audio;
+    expect(audio3.roles).toEqual(['description']);
+    expect(audio3.originalId).toBe('stream2,stream2');
+
+    const audio4 = variants[3].audio;
+    expect(audio4.roles).toEqual(['description']);
+    expect(audio4.originalId).toBe('stream2,stream2');
+  });
+
+  it('Matches streams with mismatched roles', async () => {
```

Comment:
This test simulates a similar stream to the one above, but this time with periods representing pre-roll and mid-roll ads that only have `main` audio tracks.